// ionic-angular/__ivy_ngcc__/es2015/util/ng-module-loader.js
import { Compiler, Injectable } from '@angular/core';
/**
 * NgModuleFactoryLoader that uses SystemJS to load NgModuleFactory
 */
import * as ɵngcc0 from '@angular/core';
export class NgModuleLoader {
    constructor(_compiler) {
        this._compiler = _compiler;
    }
    load(modulePath, ngModuleExport) {
        const offlineMode = this._compiler instanceof Compiler;
        return offlineMode ? loadPrecompiledFactory(modulePath, ngModuleExport) : loadAndCompile(this._compiler, modulePath, ngModuleExport);
    }
}
NgModuleLoader.ɵfac = function NgModuleLoader_Factory(t) { return new (t || NgModuleLoader)(ɵngcc0.ɵɵinject(ɵngcc0.Compiler)); };
NgModuleLoader.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: NgModuleLoader, factory: NgModuleLoader.ɵfac });
/** @nocollapse */
NgModuleLoader.ctorParameters = () => [
    { type: Compiler, },
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgModuleLoader, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc0.Compiler }]; }, null); })();
function loadAndCompile(compiler, modulePath, ngModuleExport) {
    if (!ngModuleExport) {
        ngModuleExport = 'default';
    }
    return System.import(modulePath)
        .then((rawModule) => {
        const module = rawModule[ngModuleExport];
        if (!module) {
            throw new Error(`Module ${modulePath} does not export ${ngModuleExport}`);
        }
        return compiler.compileModuleAsync(module);
    });
}
function loadPrecompiledFactory(modulePath, ngModuleExport) {
    return System.import(modulePath)
        .then((rawModule) => {
        const ngModuleFactory = rawModule[ngModuleExport];
        if (!ngModuleFactory) {
            throw new Error(`Module ${modulePath} does not export ${ngModuleExport}`);
        }
        return ngModuleFactory;
    });
}
