// ionic-angular/__ivy_ngcc__/es2015/components/toolbar/navbar.js
import { Component, ElementRef, Input, Optional, Renderer } from '@angular/core';
import { App } from '../app/app';
import { Config } from '../../config/config';
import { isTrueProperty } from '../../util/util';
import { NavController } from '../../navigation/nav-controller';
import { ToolbarBase } from './toolbar-base';
import { ViewController } from '../../navigation/view-controller';
/**
 * @name Navbar
 * @description
 * Navbar acts as the navigational toolbar, which also comes with a back
 * button. A navbar can contain a `ion-title`, any number of buttons,
 * a segment, or a searchbar. Navbars must be placed within an
 * `<ion-header>` in order for them to be placed above the content.
 * It's important to note that navbar's are part of the dynamic navigation
 * stack. If you need a static toolbar, use ion-toolbar.
 *
 * @usage
 * ```html
 * <ion-header>
 *
 *   <ion-navbar>
 *     <button ion-button icon-only menuToggle>
 *       <ion-icon name="menu"></ion-icon>
 *     </button>
 *
 *     <ion-title>
 *       Page Title
 *     </ion-title>
 *
 *     <ion-buttons end>
 *       <button ion-button icon-only (click)="openModal()">
 *         <ion-icon name="options"></ion-icon>
 *       </button>
 *     </ion-buttons>
 *   </ion-navbar>
 *
 * </ion-header>
 * ```
 *
 * @demo /docs/demos/src/navbar/
 * @see {@link ../../toolbar/Toolbar/ Toolbar API Docs}
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../app/app';
import * as ɵngcc2 from '../../navigation/view-controller';
import * as ɵngcc3 from '../../navigation/nav-controller';
import * as ɵngcc4 from '../../config/config';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../button/button';
import * as ɵngcc7 from '../icon/icon';

const _c0 = [[["", "menuToggle", ""], ["ion-buttons", "left", ""]], [["ion-buttons", "start", ""]], [["ion-buttons", "end", ""], ["ion-buttons", "right", ""]], "*"];
const _c1 = ["[menuToggle],ion-buttons[left]", "ion-buttons[start]", "ion-buttons[end],ion-buttons[right]", "*"];
export class Navbar extends ToolbarBase {
    constructor(_app, viewCtrl, navCtrl, config, elementRef, renderer) {
        super(config, elementRef, renderer);
        this._app = _app;
        this.navCtrl = navCtrl;
        /**
         * @hidden
         */
        this._hidden = false;
        /**
         * @hidden
         */
        this._hideBb = false;
        viewCtrl && viewCtrl._setNavbar(this);
        this._bbIcon = config.get('backButtonIcon');
        this._sbPadding = config.getBoolean('statusbarPadding');
        this._backText = config.get('backButtonText', 'Back');
    }
    /**
     * @input {boolean} If true, the back button will be hidden.
     */
    get hideBackButton() {
        return this._hideBb;
    }
    set hideBackButton(val) {
        this._hideBb = isTrueProperty(val);
    }
    backButtonClick(ev) {
        ev.preventDefault();
        ev.stopPropagation();
        this.navCtrl && this.navCtrl.pop(null, null);
    }
    /**
     * Set the text of the Back Button in the Nav Bar. Defaults to "Back".
     */
    setBackButtonText(text) {
        this._backText = text;
    }
    /**
     * @hidden
     */
    didEnter() {
        try {
            this._app.setTitle(this.getTitleText());
        }
        catch (e) {
            console.error(e);
        }
    }
    /**
     * @hidden
     */
    setHidden(isHidden) {
        // used to display none/block the navbar
        this._hidden = isHidden;
    }
}
Navbar.ɵfac = function Navbar_Factory(t) { return new (t || Navbar)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.App), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ViewController, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.NavController, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.Config), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(Renderer)); };
Navbar.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: Navbar, selectors: [["ion-navbar"]], hostAttrs: [1, "toolbar"], hostVars: 3, hostBindings: function Navbar_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("hidden", ctx._hidden);
        ɵngcc0.ɵɵclassProp("statusbar-padding", ctx._sbPadding);
    } }, inputs: { hideBackButton: "hideBackButton" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c1, decls: 10, vars: 8, consts: [[1, "toolbar-background", 3, "ngClass"], ["ion-button", "bar-button", 1, "back-button", 3, "ngClass", "hidden", "click"], [1, "back-button-icon", 3, "ngClass", "name"], [1, "back-button-text", 3, "ngClass"], [1, "toolbar-content", 3, "ngClass"]], template: function Navbar_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelement(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵlistener("click", function Navbar_Template_button_click_1_listener($event) { return ctx.backButtonClick($event); });
        ɵngcc0.ɵɵelement(2, "ion-icon", 2);
        ɵngcc0.ɵɵelementStart(3, "span", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵelementEnd()();
        ɵngcc0.ɵɵprojection(5);
        ɵngcc0.ɵɵprojection(6, 1);
        ɵngcc0.ɵɵprojection(7, 2);
        ɵngcc0.ɵɵelementStart(8, "div", 4);
        ɵngcc0.ɵɵprojection(9, 3);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", "toolbar-background-" + ctx._mode);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", "back-button-" + ctx._mode)("hidden", ctx._hideBb);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", "back-button-icon-" + ctx._mode)("name", ctx._bbIcon);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", "back-button-text-" + ctx._mode);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx._backText);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngClass", "toolbar-content-" + ctx._mode);
    } }, dependencies: [ɵngcc5.NgClass, ɵngcc6.Button, ɵngcc7.Icon], encapsulation: 2 });
/** @nocollapse */
Navbar.ctorParameters = () => [
    { type: App, },
    { type: ViewController, decorators: [{ type: Optional },] },
    { type: NavController, decorators: [{ type: Optional },] },
    { type: Config, },
    { type: ElementRef, },
    { type: Renderer, },
];
Navbar.propDecorators = {
    'hideBackButton': [{ type: Input },],
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Navbar, [{
        type: Component,
        args: [{
                selector: 'ion-navbar',
                template: '<div class="toolbar-background" [ngClass]="\'toolbar-background-\' + _mode"></div>' +
                    '<button (click)="backButtonClick($event)" ion-button="bar-button" class="back-button" [ngClass]="\'back-button-\' + _mode" [hidden]="_hideBb">' +
                    '<ion-icon class="back-button-icon" [ngClass]="\'back-button-icon-\' + _mode" [name]="_bbIcon"></ion-icon>' +
                    '<span class="back-button-text" [ngClass]="\'back-button-text-\' + _mode">{{_backText}}</span>' +
                    '</button>' +
                    '<ng-content select="[menuToggle],ion-buttons[left]"></ng-content>' +
                    '<ng-content select="ion-buttons[start]"></ng-content>' +
                    '<ng-content select="ion-buttons[end],ion-buttons[right]"></ng-content>' +
                    '<div class="toolbar-content" [ngClass]="\'toolbar-content-\' + _mode">' +
                    '<ng-content></ng-content>' +
                    '</div>',
                host: {
                    '[hidden]': '_hidden',
                    'class': 'toolbar',
                    '[class.statusbar-padding]': '_sbPadding'
                }
            }]
    }], function () { return [{ type: ɵngcc1.App }, { type: ɵngcc2.ViewController, decorators: [{
                type: Optional
            }] }, { type: ɵngcc3.NavController, decorators: [{
                type: Optional
            }] }, { type: ɵngcc4.Config }, { type: ɵngcc0.ElementRef }, { type: Renderer }]; }, { hideBackButton: [{
            type: Input
        }] }); })();
