// ionic-angular/__ivy_ngcc__/es2015/components/tabs/tab-highlight.js
import { Directive, ElementRef } from '@angular/core';
import { DomController } from '../../platform/dom-controller';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../platform/dom-controller';
export class TabHighlight {
    constructor(_elementRef, _dom) {
        this._elementRef = _elementRef;
        this._dom = _dom;
    }
    select(tab) {
        if (!tab) {
            return;
        }
        const dom = this._dom;
        dom.read(() => {
            const btnEle = tab.btn.getNativeElement();
            const transform = `translate3d(${btnEle.offsetLeft}px,0,0) scaleX(${btnEle.offsetWidth})`;
            dom.write(() => {
                const ele = this._elementRef.nativeElement;
                ele.style[dom.plt.Css.transform] = transform;
                if (!this._init) {
                    this._init = true;
                    dom.write(() => {
                        ele.classList.add('animate');
                    }, 80);
                }
            });
        }, 32);
    }
}
TabHighlight.ɵfac = function TabHighlight_Factory(t) { return new (t || TabHighlight)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DomController)); };
TabHighlight.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: TabHighlight, selectors: [["", 8, "tab-highlight"]] });
/** @nocollapse */
TabHighlight.ctorParameters = () => [
    { type: ElementRef, },
    { type: DomController, },
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TabHighlight, [{
        type: Directive,
        args: [{
                selector: '.tab-highlight'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.DomController }]; }, null); })();
