// ionic-angular/__ivy_ngcc__/es2015/components/nav/nav-pop-anchor.js
import { Directive, Optional } from '@angular/core';
import { DeepLinker } from '../../navigation/deep-linker';
import { ViewController } from '../../navigation/view-controller';
import { NavPop } from './nav-pop';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './nav-pop';
import * as ɵngcc2 from '../../navigation/deep-linker';
import * as ɵngcc3 from '../../navigation/view-controller';
export class NavPopAnchor {
    constructor(host, linker, viewCtrl) {
        this.host = host;
        this.linker = linker;
        this.viewCtrl = viewCtrl;
    }
    updateHref() {
        if (this.host && this.viewCtrl) {
            const previousView = this.host._nav.getPrevious(this.viewCtrl);
            this._href = (previousView && this.linker.createUrl(this.host._nav, this.viewCtrl.component, this.viewCtrl.data)) || '#';
        }
        else {
            this._href = '#';
        }
    }
    ngAfterContentInit() {
        this.updateHref();
    }
}
NavPopAnchor.ɵfac = function NavPopAnchor_Factory(t) { return new (t || NavPopAnchor)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NavPop, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DeepLinker), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ViewController, 8)); };
NavPopAnchor.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: NavPopAnchor, selectors: [["a", "navPop", ""]], hostVars: 1, hostBindings: function NavPopAnchor_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("href", ctx._href, ɵngcc0.ɵɵsanitizeUrl);
    } } });
/** @nocollapse */
NavPopAnchor.ctorParameters = () => [
    { type: NavPop, decorators: [{ type: Optional },] },
    { type: DeepLinker, },
    { type: ViewController, decorators: [{ type: Optional },] },
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NavPopAnchor, [{
        type: Directive,
        args: [{
                selector: 'a[navPop]',
                host: {
                    '[attr.href]': '_href'
                }
            }]
    }], function () { return [{ type: ɵngcc1.NavPop, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.DeepLinker }, { type: ɵngcc3.ViewController, decorators: [{
                type: Optional
            }] }]; }, null); })();
