// ionic-angular/__ivy_ngcc__/es2015/components/item/item-options.js
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { Platform } from '../../platform/platform';
import { isRightSide } from '../../util/util';
/**
 * @name ItemOptions
 * @description
 * The option buttons for an `ion-item-sliding`. These buttons can be placed either on the left or right side.
 * You can combine the `(ionSwipe)` event plus the `expandable` directive to create a full swipe action for the item.
 *
 * @usage
 *
 * ```html
 * <ion-item-sliding>
 *   <ion-item>
 *     Item 1
 *   </ion-item>
 *   <ion-item-options side="right" (ionSwipe)="saveItem(item)">
 *     <button ion-button expandable (click)="saveItem(item)">
 *       <ion-icon name="star"></ion-icon>
 *     </button>
 *   </ion-item-options>
 * </ion-item-sliding>
 *```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../platform/platform';
export class ItemOptions {
    constructor(_elementRef, _plt) {
        this._elementRef = _elementRef;
        this._plt = _plt;
        /**
         * @output {event} Emitted when the item has been fully swiped.
         */
        this.ionSwipe = new EventEmitter();
    }
    /**
     * @hidden
     */
    isRightSide() {
        return isRightSide(this.side, this._plt.isRTL, true);
    }
    /**
     * @hidden
     */
    width() {
        return this._elementRef.nativeElement.offsetWidth;
    }
}
ItemOptions.ɵfac = function ItemOptions_Factory(t) { return new (t || ItemOptions)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Platform)); };
ItemOptions.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: ItemOptions, selectors: [["ion-item-options"]], inputs: { side: "side" }, outputs: { ionSwipe: "ionSwipe" } });
/** @nocollapse */
ItemOptions.ctorParameters = () => [
    { type: ElementRef, },
    { type: Platform, },
];
ItemOptions.propDecorators = {
    'side': [{ type: Input },],
    'ionSwipe': [{ type: Output },],
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ItemOptions, [{
        type: Directive,
        args: [{
                selector: 'ion-item-options'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.Platform }]; }, { ionSwipe: [{
            type: Output
        }], side: [{
            type: Input
        }] }); })();
