// ionic-angular/__ivy_ngcc__/es2015/components/item/item-divider.js
import { Directive, ElementRef, Renderer } from '@angular/core';
import { Config } from '../../config/config';
import { Ion } from '../ion';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../config/config';
export class ItemDivider extends Ion {
    constructor(config, elementRef, renderer) {
        super(config, elementRef, renderer, 'item-divider');
    }
}
ItemDivider.ɵfac = function ItemDivider_Factory(t) { return new (t || ItemDivider)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Config), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(Renderer)); };
ItemDivider.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: ItemDivider, selectors: [["ion-item-divider"]], hostAttrs: [1, "item-divider"], features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
/** @nocollapse */
ItemDivider.ctorParameters = () => [
    { type: Config, },
    { type: ElementRef, },
    { type: Renderer, },
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ItemDivider, [{
        type: Directive,
        args: [{
                selector: 'ion-item-divider',
                host: {
                    'class': 'item-divider'
                }
            }]
    }], function () { return [{ type: ɵngcc1.Config }, { type: ɵngcc0.ElementRef }, { type: Renderer }]; }, null); })();
