// ionic-angular/__ivy_ngcc__/es2015/components/app/overlay-portal.js
import { ComponentFactoryResolver, Directive, ElementRef, ErrorHandler, Inject, Input, NgZone, Optional, Renderer, ViewContainerRef, forwardRef } from '@angular/core';
import { App } from './app';
import { Config } from '../../config/config';
import { DeepLinker } from '../../navigation/deep-linker';
import { DomController } from '../../platform/dom-controller';
import { GestureController } from '../../gestures/gesture-controller';
import { NavControllerBase } from '../../navigation/nav-controller-base';
import { Platform } from '../../platform/platform';
import { TransitionController } from '../../transitions/transition-controller';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../config/config';
import * as ɵngcc2 from '../../platform/platform';
import * as ɵngcc3 from '../../gestures/gesture-controller';
import * as ɵngcc4 from '../../transitions/transition-controller';
import * as ɵngcc5 from '../../navigation/deep-linker';
import * as ɵngcc6 from '../../platform/dom-controller';
import * as ɵngcc7 from './app';
export class OverlayPortal extends NavControllerBase {
    constructor(app, config, plt, elementRef, zone, renderer, cfr, gestureCtrl, transCtrl, linker, viewPort, domCtrl, errHandler) {
        super(null, app, config, plt, elementRef, zone, renderer, cfr, gestureCtrl, transCtrl, linker, domCtrl, errHandler);
        this._isPortal = true;
        this._init = true;
        this.setViewport(viewPort);
        // on every page change make sure the portal has
        // dismissed any views that should be auto dismissed on page change
        app.viewDidLeave.subscribe((view) => {
            if (!view.isOverlay) {
                this.dismissPageChangeViews();
            }
        });
    }
    set _overlayPortal(val) {
        this._zIndexOffset = (val || 0);
    }
    ngOnDestroy() {
        this.destroy();
    }
    /*
     * @private
     */
    getType() {
        return 'portal';
    }
    /*
     * @private
     */
    getSecondaryIdentifier() {
        return null;
    }
}
OverlayPortal.ɵfac = function OverlayPortal_Factory(t) { return new (t || OverlayPortal)(ɵngcc0.ɵɵdirectiveInject(forwardRef(() => App)), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Config), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Platform), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(Renderer), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.GestureController), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TransitionController), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.DeepLinker, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.DomController), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ErrorHandler)); };
OverlayPortal.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: OverlayPortal, selectors: [["", "overlay-portal", ""]], inputs: { _overlayPortal: ["overlay-portal", "_overlayPortal"] }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
/** @nocollapse */
OverlayPortal.ctorParameters = () => [
    { type: App, decorators: [{ type: Inject, args: [forwardRef(() => App),] },] },
    { type: Config, },
    { type: Platform, },
    { type: ElementRef, },
    { type: NgZone, },
    { type: Renderer, },
    { type: ComponentFactoryResolver, },
    { type: GestureController, },
    { type: TransitionController, },
    { type: DeepLinker, decorators: [{ type: Optional },] },
    { type: ViewContainerRef, },
    { type: DomController, },
    { type: ErrorHandler, },
];
OverlayPortal.propDecorators = {
    '_overlayPortal': [{ type: Input, args: ['overlay-portal',] },],
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OverlayPortal, [{
        type: Directive,
        args: [{
                selector: '[overlay-portal]'
            }]
    }], function () { return [{ type: ɵngcc7.App, decorators: [{
                type: Inject,
                args: [forwardRef(() => App)]
            }] }, { type: ɵngcc1.Config }, { type: ɵngcc2.Platform }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.NgZone }, { type: Renderer }, { type: ɵngcc0.ComponentFactoryResolver }, { type: ɵngcc3.GestureController }, { type: ɵngcc4.TransitionController }, { type: ɵngcc5.DeepLinker, decorators: [{
                type: Optional
            }] }, { type: ɵngcc0.ViewContainerRef }, { type: ɵngcc6.DomController }, { type: ɵngcc0.ErrorHandler }]; }, { _overlayPortal: [{
            type: Input,
            args: ['overlay-portal']
        }] }); })();
