// ionic-angular/__ivy_ngcc__/es2015/components/menu/menu-close.js
import { Directive, HostListener, Input } from '@angular/core';
import { MenuController } from '../app/menu-controller';
/**
 * @name MenuClose
 * @description
 * The `menuClose` directive can be placed on any button to close an open menu.
 *
 * @usage
 *
 * A simple `menuClose` button can be added using the following markup:
 *
 * ```html
 * <button ion-button menuClose>Close Menu</button>
 * ```
 *
 * To close a certain menu by its id or side, give the `menuClose`
 * directive a value.
 *
 * ```html
 * <button ion-button menuClose="left">Close Left Menu</button>
 * ```
 *
 * @demo /docs/demos/src/menu/
 * @see {@link /docs/components#menus Menu Component Docs}
 * @see {@link ../../menu/Menu Menu API Docs}
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../app/menu-controller';
export class MenuClose {
    constructor(_menu) {
        this._menu = _menu;
    }
    /**
    * @hidden
    */
    close() {
        const menu = this._menu.get(this.menuClose);
        menu && menu.close();
    }
}
MenuClose.ɵfac = function MenuClose_Factory(t) { return new (t || MenuClose)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MenuController)); };
MenuClose.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: MenuClose, selectors: [["", "menuClose", ""]], hostBindings: function MenuClose_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function MenuClose_click_HostBindingHandler() { return ctx.close(); });
    } }, inputs: { menuClose: "menuClose" } });
/** @nocollapse */
MenuClose.ctorParameters = () => [
    { type: MenuController, },
];
MenuClose.propDecorators = {
    'menuClose': [{ type: Input },],
    'close': [{ type: HostListener, args: ['click',] },],
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuClose, [{
        type: Directive,
        args: [{
                selector: '[menuClose]'
            }]
    }], function () { return [{ type: ɵngcc1.MenuController }]; }, { 
    /**
    * @hidden
    */
    close: [{
            type: HostListener,
            args: ['click']
        }], menuClose: [{
            type: Input
        }] }); })();
