// ionic-angular/__ivy_ngcc__/es2015/components/segment/segment-button.js
import { Component, EventEmitter, HostListener, Input, Output, ViewEncapsulation } from '@angular/core';
import { isPresent, isTrueProperty } from '../../util/util';
/**
 * @name SegmentButton
 * @description
 * The child buttons of the `ion-segment` component. Each `ion-segment-button` must have a value.
 *
 * @usage
 *
 * ```html
 * <ion-content>
 *   <!-- Segment buttons with icons -->
 *   <ion-segment [(ngModel)]="icons" color="secondary">
 *     <ion-segment-button value="camera">
 *       <ion-icon name="camera"></ion-icon>
 *     </ion-segment-button>
 *     <ion-segment-button value="bookmark">
 *       <ion-icon name="bookmark"></ion-icon>
 *     </ion-segment-button>
 *   </ion-segment>
 *
 *   <!-- Segment buttons with text -->
 *   <ion-segment [(ngModel)]="relationship" color="primary">
 *     <ion-segment-button value="friends" (ionSelect)="selectedFriends()">
 *       Friends
 *     </ion-segment-button>
 *     <ion-segment-button value="enemies" (ionSelect)="selectedEnemies()">
 *       Enemies
 *     </ion-segment-button>
 *   </ion-segment>
 * </ion-content>
 * ```
 *
 *
 * @demo /docs/demos/src/segment/
 * @see {@link /docs/components#segment Segment Component Docs}
 * @see {@link /docs/api/components/segment/Segment/ Segment API Docs}
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class SegmentButton {
    constructor() {
        this.isActive = false;
        this._disabled = false;
        /**
         * @output {SegmentButton} Emitted when a segment button has been clicked.
         */
        this.ionSelect = new EventEmitter();
    }
    /**
     * @input {boolean} If true, the user cannot interact with this element.
     */
    get disabled() {
        return this._disabled;
    }
    set disabled(val) {
        this._disabled = isTrueProperty(val);
    }
    /**
     * @hidden
     * On click of a SegmentButton
     */
    onClick() {
        (void 0) /* console.debug */;
        this.ionSelect.emit(this);
    }
    /**
     * @hidden
     */
    ngOnInit() {
        if (!isPresent(this.value)) {
            console.warn('<ion-segment-button> requires a "value" attribute');
        }
    }
}
SegmentButton.ɵfac = function SegmentButton_Factory(t) { return new (t || SegmentButton)(); };
SegmentButton.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SegmentButton, selectors: [["ion-segment-button"]], hostAttrs: ["tappable", "", "role", "button", 1, "segment-button"], hostVars: 5, hostBindings: function SegmentButton_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function SegmentButton_click_HostBindingHandler() { return ctx.onClick(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-pressed", ctx.isActive);
        ɵngcc0.ɵɵclassProp("segment-button-disabled", ctx._disabled)("segment-activated", ctx.isActive);
    } }, inputs: { disabled: "disabled", value: "value" }, outputs: { ionSelect: "ionSelect" }, ngContentSelectors: _c0, decls: 2, vars: 0, consts: [[1, "button-effect"]], template: function SegmentButton_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelement(1, "div", 0);
    } }, encapsulation: 2 });
/** @nocollapse */
SegmentButton.ctorParameters = () => [];
SegmentButton.propDecorators = {
    'value': [{ type: Input },],
    'ionSelect': [{ type: Output },],
    'disabled': [{ type: Input },],
    'onClick': [{ type: HostListener, args: ['click',] },],
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SegmentButton, [{
        type: Component,
        args: [{
                selector: 'ion-segment-button',
                template: '<ng-content></ng-content>' +
                    '<div class="button-effect"></div>',
                host: {
                    'tappable': '',
                    'class': 'segment-button',
                    'role': 'button',
                    '[class.segment-button-disabled]': '_disabled',
                    '[class.segment-activated]': 'isActive',
                    '[attr.aria-pressed]': 'isActive'
                },
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return []; }, { ionSelect: [{
            type: Output
        }], disabled: [{
            type: Input
        }], 
    /**
     * @hidden
     * On click of a SegmentButton
     */
    onClick: [{
            type: HostListener,
            args: ['click']
        }], value: [{
            type: Input
        }] }); })();
