// ionic-angular/__ivy_ngcc__/es2015/components/show-hide-when/show-when.js
import { Attribute, Directive, NgZone } from '@angular/core';
import { DisplayWhen } from './display-when';
import { Platform } from '../../platform/platform';
/**
 *
 * @name ShowWhen
 * @description
 * The `showWhen` attribute takes a string that represents a platform or screen orientation.
 * The element the attribute is added to will only be shown when that platform or screen orientation is active.
 *
 * Complements the [hideWhen attribute](../HideWhen). If the `showWhen` attribute is used on an
 * element that also has the `hideWhen` attribute, the element will not show if `hideWhen` evaluates
 * to `true` or `showWhen` evaluates to `false`. If the `hidden` attribute is also added, the element
 * will not show if `hidden` evaluates to `true`.
 *
 * View the [Platform API docs](../../../platform/Platform) for more information on the different
 * platforms you can use.
 *
 * @usage
 * ```html
 * <div showWhen="android">
 *  I am visible on Android!
 * </div>
 *
 * <div showWhen="ios">
 *  I am visible on iOS!
 * </div>
 *
 * <div showWhen="android,ios">
 *  I am visible on Android and iOS!
 * </div>
 *
 * <div showWhen="portrait">
 *  I am visible on Portrait!
 * </div>
 *
 * <div showWhen="landscape">
 *  I am visible on Landscape!
 * </div>
 * ```
 * @demo /docs/demos/src/show-when/
 * @see {@link ../HideWhen HideWhen API Docs}
 * @see {@link ../../../platform/Platform Platform API Docs}
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../platform/platform';
export class ShowWhen extends DisplayWhen {
    constructor(showWhen, plt, zone) {
        super(showWhen, plt, zone);
    }
}
ShowWhen.ɵfac = function ShowWhen_Factory(t) { return new (t || ShowWhen)(ɵngcc0.ɵɵinjectAttribute('showWhen'), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Platform), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone)); };
ShowWhen.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: ShowWhen, selectors: [["", "showWhen", ""]], hostVars: 2, hostBindings: function ShowWhen_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("hidden-show-when", !ctx.isMatch);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
/** @nocollapse */
ShowWhen.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Attribute, args: ['showWhen',] },] },
    { type: Platform, },
    { type: NgZone, },
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ShowWhen, [{
        type: Directive,
        args: [{
                selector: '[showWhen]',
                host: {
                    '[class.hidden-show-when]': '!isMatch'
                }
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Attribute,
                args: ['showWhen']
            }] }, { type: ɵngcc1.Platform }, { type: ɵngcc0.NgZone }]; }, null); })();
