// ionic-angular/__ivy_ngcc__/es2015/components/option/option.js
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { isPresent, isTrueProperty } from '../../util/util';
/**
 * @name Option
 * @description
 * `ion-option` is a child component of `ion-select`. Similar to the native option element, `ion-option` can take a value and a selected property.
 *
 * @demo /docs/demos/src/select/
 */
import * as ɵngcc0 from '@angular/core';
export class Option {
    constructor(_elementRef) {
        this._elementRef = _elementRef;
        this._selected = false;
        this._disabled = false;
        /**
         * @output {any} Event to evaluate when option is selected.
         */
        this.ionSelect = new EventEmitter();
    }
    /**
     * @input {boolean} If true, the user cannot interact with this element.
     */
    get disabled() {
        return this._disabled;
    }
    set disabled(val) {
        this._disabled = isTrueProperty(val);
    }
    /**
     * @input {boolean} If true, the element is selected.
     */
    get selected() {
        return this._selected;
    }
    set selected(val) {
        this._selected = isTrueProperty(val);
    }
    /**
     * @input {any} The value of the option.
     */
    get value() {
        if (isPresent(this._value)) {
            return this._value;
        }
        return this.text;
    }
    set value(val) {
        this._value = val;
    }
    /**
     * @hidden
     */
    get text() {
        return this._elementRef.nativeElement.textContent;
    }
}
Option.ɵfac = function Option_Factory(t) { return new (t || Option)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
Option.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: Option, selectors: [["ion-option"]], inputs: { disabled: "disabled", selected: "selected", value: "value" }, outputs: { ionSelect: "ionSelect" } });
/** @nocollapse */
Option.ctorParameters = () => [
    { type: ElementRef, },
];
Option.propDecorators = {
    'disabled': [{ type: Input },],
    'selected': [{ type: Input },],
    'value': [{ type: Input },],
    'ionSelect': [{ type: Output },],
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Option, [{
        type: Directive,
        args: [{
                selector: 'ion-option'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { ionSelect: [{
            type: Output
        }], disabled: [{
            type: Input
        }], selected: [{
            type: Input
        }], value: [{
            type: Input
        }] }); })();
