// ionic-angular/__ivy_ngcc__/es2015/components/backdrop/backdrop.js
import { Directive, ElementRef, Renderer } from '@angular/core';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
export class Backdrop {
    constructor(_elementRef, _renderer) {
        this._elementRef = _elementRef;
        this._renderer = _renderer;
    }
    getNativeElement() {
        return this._elementRef.nativeElement;
    }
    setElementClass(className, add) {
        this._renderer.setElementClass(this._elementRef.nativeElement, className, add);
    }
}
Backdrop.ɵfac = function Backdrop_Factory(t) { return new (t || Backdrop)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(Renderer)); };
Backdrop.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: Backdrop, selectors: [["ion-backdrop"]], hostAttrs: ["role", "presentation", "tappable", "", "disable-activated", ""] });
/** @nocollapse */
Backdrop.ctorParameters = () => [
    { type: ElementRef, },
    { type: Renderer, },
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Backdrop, [{
        type: Directive,
        args: [{
                selector: 'ion-backdrop',
                host: {
                    'role': 'presentation',
                    'tappable': '',
                    'disable-activated': ''
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: Renderer }]; }, null); })();
